File: julia/gaussian.jl
include("prox.jl")

using LinearAlgebra
using SparseArrays
using Base.Threads
using Distributions

function deviance(::Normal, residual, y)
    norm(residual)^2
end

function primal(::Normal, residual, β, λ)
    0.5*norm(residual)^2 + λ*norm(β, 1)
end

function dual(::Normal, y, residual)
    dot(residual, y) - 0.5*norm(residual)^2
    # or 0.5*norm(y)^2 - 0.5*norm(X*β)^2
end

# function residual(::Normal, Xβ, y)
#     y .- Xβ
# end

function scaledDual(::Normal, y, residual, λ, dual_scale)
    if dual_scale == 0
        return 0.0
    end

    α = λ/dual_scale
    α*dot(residual, y) - 0.5*(α*norm(residual))^2
end

function gradient(::Normal, X, residual)
    -X'*residual
end

function hessian(::Normal, X::Matrix{Float64}, X_mean_scaled, w, indices, standardize)
    Symmetric(X[:, indices]'X[:, indices])
end

function hessian(::Normal, X::SparseMatrixCSC{Float64}, X_mean_scaled, w, indices, standardize)
    H = Matrix(X[:, indices]'X[:, indices])

    if standardize
        H .-= size(X, 1)*(X_mean_scaled[indices]*X_mean_scaled[indices]')
    end

    Symmetric(H)
end

function hessianUpperrightBlock(::Normal,
                                Xa::Matrix{Float64},
                                Xb::Matrix{Float64},
                                Xa_mean_scaled,
                                Xb_mean_scaled,
                                w,
                                standardize)
    Xa'Xb
end

function hessianUpperrightBlock(::Normal,
                                Xa::SparseMatrixCSC{Float64},
                                Xb::SparseMatrixCSC{Float64},
                                Xa_mean_scaled,
                                Xb_mean_scaled,
                                w,
                                standardize)
    H = Matrix(Xa'Xb)

    if standardize
        H .-= size(Xa, 1)*(Xa_mean_scaled*Xb_mean_scaled')
    end

    H
end

function updateResidual!(::Normal,
                         residual::Vector{T},
                         y::Vector{T},
                         X::Matrix{T},
                         β::Vector{T},
                         X_mean_scaled::Vector{T},
                         indices,
                         standardize::Bool) where {T <: Float64}
    residual .= y .- X[:, indices]*β[indices]
end

function updateResidual!(::Normal,
                         residual::Vector{T},
                         y::Vector{T},
                         X::SparseMatrixCSC{T, Int64},
                         β::Vector{T},
                         X_mean_scaled::Vector{T},
                         indices,
                         standardize::Bool) where {T <: Float64}

    residual .= y .- Array(X[:, indices]*β[indices])

    if standardize
        residual .+= dot(β[indices], X_mean_scaled[indices])
    end
end

function fitNullModel(family::Normal, y, intercept)
    mean(y)
end

function safeScreening!(::Normal,
                        X,
                        X_mean_scaled,
                        residual,
                        c,
                        XTcenter,
                        β,
                        screened,
                        X²sums,
                        r_screen,
                        standardize)

    for j in findall(screened)
        r_normX_j = r_screen * sqrt(X²sums[j])

        if r_normX_j >= 1
            continue
        end

        if abs(XTcenter[j]) + r_normX_j < 1
            if β[j] != 0
                # update residual
                residual .+= β[j] * X[:, j]
                if issparse(X) && standardize
                    residual .-= X_mean_scaled[j] * β[j]
                end
                β[j] = 0
            end

            screened[j] = false
        end
    end
end

function lasso(family::Normal,
               X::AbstractMatrix{T},
               X²sums::Vector{T},
               X_mean_scaled::Vector{T},
               y::Vector{T},
               β::Vector{T},
               c::Vector{T},
               residual::Vector{T},
               λ::T,
               λ_max::T,
               screened::BitVector,
               null_deviance::Float64;
               standardize::Bool = true,
               screening_type = "hessian",
               maxit::Int64 = Int64(1e6),
               tol_decr::T = 1e-7,
               tol_gap::T = 1e-4,
               tol_infeas::T = 1e-4,
               verbosity::Int = 0) where {T <: Float64}

    n, p = size(X)

    it = 0

    X_is_sparse = issparse(X)

    primal_value = 0.0
    dual_value = 0.0

    XTcenter = Vector{Float64}(undef, p)
    dual_scale = λ

    screen_interval = 10

    if screening_type in ["gap_safe"]
        screened .= true
    end

    n_screened = 0
    screened_set = findall(screened)

    if !isempty(screened_set)
        updateResidual!(family, residual, y, X, β, X_mean_scaled, screened_set, standardize)

        while it < maxit
            it += 1

            if verbosity >= 2
                println("\t\titer: $(it)")
            end

            if screening_type in ["gap_safe"] && (mod(it, screen_interval) == 0 || it == 1)
                if it > 1
                    # need to update correlation for all screened predictors
                    updateResidual!(family,
                                    residual,
                                    y,
                                    X,
                                    β,
                                    X_mean_scaled,
                                    screened_set,
                                    standardize)
                end
                updateCorrelation!(c, X, X_mean_scaled, residual, screened_set, standardize)

                dual_scale = max(λ, maximum(abs.(c)))

                primal_value = primal(family, residual, β[screened_set], λ)
                dual_value = scaledDual(family, y, residual, λ, dual_scale)
                duality_gap = abs(primal_value - dual_value)

                max_infeasibility = if λ > 0 maximum(abs.(c[screened_set]) .- λ) else 0.0 end

                if duality_gap <= tol_gap*null_deviance && max_infeasibility <= tol_infeas*λ_max
                    break
                end

                if screening_type == "gap_safe"
                    for j in 1:p
                        XTcenter[j] = c[j] / dual_scale
                    end
                    r_screen = sqrt(duality_gap) / λ
                end

                safeScreening!(Normal(),
                              X,
                              X_mean_scaled,
                              residual,
                              c,
                              XTcenter,
                              β,
                              screened,
                              X²sums,
                              r_screen,
                              standardize)
                screened_set = findall(screened)
            end

            n_screened += length(screened_set)

            primal_value_old = primal(family, residual, β[screened_set], λ)

            for j in screened_set
                β_j_old = β[j]
                c[j] = dot(X[:, j], residual)
                if standardize && issparse(X)
                    c[j] -= X_mean_scaled[j]*sum(residual)
                end
                β[j] = prox(c[j] + X²sums[j]*β_j_old, λ)/X²sums[j]

                if β_j_old != β[j]
                    βdiff = β[j] - β_j_old
                    residual .-= X[:, j] * βdiff

                    if X_is_sparse && standardize
                        residual .+= X_mean_scaled[j] * βdiff
                    end
                end
            end

            primal_value = primal(family, residual, β[screened_set], λ)
            primal_value_change = primal_value - primal_value_old

            if verbosity >= 2
                println("\t\tprimal_value_change=$(primal_value_change)")
            end

            if abs(primal_value_change) <= tol_gap*primal_value
                dual_value = dual(family, y, residual)
                duality_gap = abs(primal_value - dual_value)

                updateCorrelation!(c, X, X_mean_scaled, residual, screened_set, standardize)

                max_infeasibility = if λ > 0 maximum(abs.(c[screened_set]) .- λ) else 0.0 end

                if verbosity >= 2
                    println("\t\tmax_infeas: $(max_infeasibility/λ_max)")
                    println("\t\tduality_gap: $(duality_gap/null_deviance)")
                end

                if max_infeasibility <= tol_infeas*λ_max && duality_gap <= tol_gap*null_deviance
                    break
                end
            end
        end
    else
        fill!(β, 0)
    end

    avg_screened = n_screened/it

    (passes = it,
     primal = primal_value,
     dual = dual_value,
     dual_scale = dual_scale,
     avg_screened = avg_screened)
end

File: src/gaussian.cpp
#include "gaussian.h"
#include "prox.h"
#include "utils.h"

Gaussian::Gaussian(const std::string family,
                   arma::vec& y,
                   arma::vec& beta,
                   arma::vec& residual,
                   arma::vec& Xbeta,
                   arma::vec& c,
                   const arma::vec& X_mean_scaled,
                   const arma::vec& X_norms_squared,
                   const arma::uword n,
                   const arma::uword p,
                   const bool standardize)
  : Model{ family,          y, beta, residual,   Xbeta, c, X_mean_scaled,
           X_norms_squared, n, p,    standardize }
{}

double
Gaussian::primal(const double lambda)
{
  return 0.5 * std::pow(norm(residual), 2) + lambda * norm(beta, 1);
}

double
Gaussian::primal(const double lambda, const arma::uvec& screened_set)
{
  return 0.5 * std::pow(arma::norm(residual), 2) +
         lambda * arma::norm(beta(screened_set), 1);
}

double
Gaussian::dual()
{
  return arma::dot(residual, y) - 0.5 * std::pow(arma::norm(residual), 2);
}

double
Gaussian::scaledDual(const double lambda)
{
  if (dual_scale == 0) {
    return 0;
  } else {
    double alpha = lambda / dual_scale;

    return alpha * arma::dot(residual, y) -
           0.5 * std::pow(alpha * arma::norm(residual), 2);
  }
}

double
Gaussian::deviance()
{
  return std::pow(arma::norm(residual), 2);
}

double
Gaussian::hessianTerm(const arma::mat& X, const arma::uword j)
{
  return X_norms_squared(j);
}

double
Gaussian::hessianTerm(const arma::sp_mat& X, const arma::uword j)
{
  return X_norms_squared(j);
}

void
Gaussian::updateResidual()
{
  residual = y - Xbeta;
}

void
Gaussian::adjustResidual(const arma::mat& X,
                         const arma::uword j,
                         const double beta_diff)
{
  residual -= X.col(j) * beta_diff;
}

void
Gaussian::adjustResidual(const arma::sp_mat& X,
                         const arma::uword j,
                         const double beta_diff)
{
  residual -= X.col(j) * beta_diff;

  if (standardize)
    residual += X_mean_scaled(j) * beta_diff;
}

arma::mat
Gaussian::hessian(const arma::mat& X, const arma::uvec& ind)
{
  return X.cols(ind).t() * X.cols(ind);
}

arma::mat
Gaussian::hessian(const arma::sp_mat& X, const arma::uvec& ind)
{
  using namespace arma;

  mat H = conv_to<mat>::from(X.cols(ind).t() * X.cols(ind));

  if (standardize)
    H -= X.n_rows * X_mean_scaled(ind) * X_mean_scaled(ind).t();

  return H;
}

arma::mat
Gaussian::hessianUpperRight(const arma::mat& X,
                            const arma::uvec& ind_a,
                            const arma::uvec& ind_b)
{
  return X.cols(ind_a).t() * X.cols(ind_b);
}

arma::mat
Gaussian::hessianUpperRight(const arma::sp_mat& X,
                            const arma::uvec& ind_a,
                            const arma::uvec& ind_b)
{
  using namespace arma;

  mat H(ind_a.n_elem, ind_b.n_elem);

  if (ind_b.n_elem == 1) {
    uword i = 0;
    for (auto&& j : ind_a) {
      H(i, 0) = dot(X.col(j), X.col(as_scalar(ind_b)));
      i++;
    }
  } else {
    H = conv_to<mat>::from(X.cols(ind_a).t() * X.cols(ind_b));
  }

  if (standardize)
    H -= X.n_rows * X_mean_scaled(ind_a) * X_mean_scaled(ind_b).t();

  return H;
}

void
Gaussian::updateGradientOfCorrelation(arma::vec& c_grad,
                                      const arma::mat& X,
                                      const arma::vec& Hinv_s,
                                      const arma::vec& s,
                                      const arma::uvec& active_set,
                                      const arma::uvec& restricted_set)
{
  using namespace arma;

  uvec inactive_restricted = setDiff(restricted_set, active_set);

  const vec tmp = X.cols(active_set) * Hinv_s;

  c_grad.zeros();

  c_grad(inactive_restricted) = tmp.t() * X.cols(inactive_restricted);
  c_grad(active_set) = s(active_set);
}

void
Gaussian::updateGradientOfCorrelation(arma::vec& c_grad,
                                      const arma::sp_mat& X,
                                      const arma::vec& Hinv_s,
                                      const arma::vec& s,
                                      const arma::uvec& active_set,
                                      const arma::uvec& restricted_set)
{
  using namespace arma;

  uvec inactive_restricted = setDiff(restricted_set, active_set);

  c_grad.zeros();

  if (standardize) {
    vec tmp =
      X.cols(active_set) * Hinv_s - dot(X_mean_scaled(active_set), Hinv_s);

    double tmp_sum = sum(tmp);

    for (auto&& j : inactive_restricted) {
      c_grad(j) = dot(X.col(j), tmp) - X_mean_scaled(j) * tmp_sum;
    }
  } else {
    vec tmp = X.cols(active_set) * Hinv_s;

    for (auto&& j : inactive_restricted) {
      c_grad(j) = dot(X.col(j), tmp);
    }
  }

  c_grad(active_set) = s(active_set);
}

void
Gaussian::standardizeY()
{
  y -= arma::mean(y);
}

double
Gaussian::safeScreeningRadius(const double duality_gap, const double lambda)
{
  return std::sqrt(duality_gap) / lambda;
}
